### PR TITLE
Io ex

### PR DESCRIPTION
**Pull request by ItamarB-DI**

### CODE REVIEW

File: headers/Ex1.hpp [hunk 1]
```diff
@@ -1,68 +1,123 @@
+#include <memory> //unique_ptr
+
 template <typename T>
```
Comment (idoa-DI) [inline headers/Ex1.hpp]: please merge from `main`

File: IO/Ex1.hpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
+
+#include <fstream> //ios_base::openmode
+#include <string>
+
+class FileHandler {
```
Comment (idoa-DI) [inline IO/Ex1.hpp:9]: could you add unit tests for this class?

File: IO/Ex1.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
+
+#include <stdexcept> //runtime_error
+#include <cassert>
+
+#include "Ex1.hpp"
+
+/************************************* Main Functions *******************************************/
```
Comment (idoa-DI) [inline IO/Ex1.cpp]: this comment has no purpose..

File: IO/Ex1.cpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
+
+#include <stdexcept> //runtime_error
```
Comment (idoa-DI) [inline IO/Ex1.cpp]: This is the kind of comments that may be invalidate really quickly (adding usage of another symbol from this header for example).
No need for it.

File: IO/Ex1.cpp [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    m_fd = std::move(other.m_fd);
+    m_path = std::move(other.m_path);
+    m_permissions = other.m_permissions;
+
+    other.m_permissions = std::ios_base::openmode(0);
+
+    return *this;
+} 
+
+std::fstream &FileHandler::getFD() {
```
Comment (idoa-DI) [inline IO/Ex1.cpp]: fstreams are not file descriptors

File: IO/Ex1.hpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    FileHandler& operator=(const FileHandler &other); //Copy Assignment
+    FileHandler& operator=(FileHandler &&other) noexcept; //Move Assignment
+
+    std::fstream &getFD();
+    static void createFile(const std::string filename);
+
+    ~FileHandler() noexcept = default; //Dtor
+
+private:
+    std::fstream m_fd;
```
Comment (idoa-DI) [inline IO/Ex1.hpp]: fstreams are not file descriptors

File: IO/Ex1.cpp [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+/************************************* Main Functions *******************************************/
+
+//Ctor
+FileHandler::FileHandler(const std::string path, std::ios_base::openmode permissions)
+: m_path(path), 
+  m_permissions(permissions)
+{
+
+    m_fd.open(path, permissions);
+    if (!m_fd.is_open()) {
```
Comment (idoa-DI) [inline IO/Ex1.cpp]: ~You can check for failure by checking the result of `open()`~
Sorry I accidentally read the open() documentation of std::basic_filebuf

File: IO/Ex1.hpp [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+public:
+    explicit FileHandler(const std::string path, std::ios_base::openmode permissions); //Ctor
+    FileHandler(const FileHandler &other); //CCtor
+    FileHandler(FileHandler &&other) noexcept; //MCtor
+
+    FileHandler& operator=(const FileHandler &other); //Copy Assignment
+    FileHandler& operator=(FileHandler &&other) noexcept; //Move Assignment
+
+    std::fstream &getFD();
+    static void createFile(const std::string filename);
```
Comment (idoa-DI) [inline IO/Ex1.hpp]: How is this function aligned with the rest of this class? What is the relation between the current file opened in the constructor and the new file created by this function?

File: IO/Ex1.cpp [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+//Copy Assignment
+FileHandler& FileHandler::operator=(const FileHandler &other) {
+
+    assert(this != &other);
+
+    if (m_fd.is_open()) {
+        m_fd.close();
+    }
+
+    m_fd.open(other.m_path, other.m_permissions);
```
Comment (idoa-DI) [inline IO/Ex1.cpp]: `fstream` contains more metadata which is lost if you just open the path here

File: IO/Ex1.cpp [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+  m_path(std::move(other.m_path)),
+  m_permissions(other.m_permissions)
+{
+    other.m_permissions = std::ios_base::openmode(0);
+}
+
+//Copy Assignment
+FileHandler& FileHandler::operator=(const FileHandler &other) {
+
+    assert(this != &other);
```
Comment (idoa-DI) [inline IO/Ex1.cpp:51]: If you use `assert()` you need to make sure `NDEBUG` is defined in your testing build

File: IO/Ex1.hpp (hunk 6 above)
Comment (ItamarB-DI) [reply IO/Ex1.hpp]: do you mean i should replace the variable name ?  (to somthing like m_file_stream ?)

File: IO/Ex1.hpp (hunk 8 above)
Comment (ItamarB-DI) [reply IO/Ex1.hpp]: it is a static function that helps with creating files without making the user to cease control on them.
but tell me what you think if it is better to make this method private and just call it at the Ctor.

File: IO/Ex1.cpp (hunk 4 above)
Comment (ItamarB-DI) [reply IO/Ex1.cpp]: it is something that stuck with me that if i include a library i list at least 1 thing i used from there

File: IO/Ex1.hpp (hunk 8 above)
Comment (idoa-DI) [reply IO/Ex1.hpp]: no need for this.

File: IO/Ex1.cpp [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+FileHandler::FileHandler(FileHandler &&other) noexcept 
+: m_file(std::move(other.m_file)),
+  m_path(std::move(other.m_path)),
+  m_permissions(other.m_permissions)
+{
+    other.m_permissions = std::ios_base::openmode(0);
+}
+
+//Copy Assignment
+FileHandler& FileHandler::operator=(FileHandler &other) {
```
Comment (idoa-DI) [inline IO/Ex1.cpp:49]: please add a comment about why this `operator=` receives non const reference

File: IO/Ex1.cpp [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+{
+
+    m_file.open(path, permissions);
+    if (!m_file.is_open()) {
+        throw std::runtime_error("Failed to open file: " + path);
+    }
+}
+
+//CCtor
+FileHandler::FileHandler(FileHandler &other)
```
Comment (idoa-DI) [inline IO/Ex1.cpp:22]: please add a comment about why this copy consturctor receives non const reference

File: IO/Ex1.cpp [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    m_permissions = other.m_permissions;
+
+    other.m_permissions = std::ios_base::openmode(0);
+
+    return *this;
+} 
+
+std::fstream &FileHandler::getFile() {
+
+    return m_file;
```
Comment (idoa-DI) [inline IO/Ex1.cpp]: what about m_stream?

File: IO/Ex1.cpp (hunk 11 above)
Comment (ItamarB-DI) [reply IO/Ex1.cpp:49]: added in the header

File: IO/Ex1.cpp (hunk 12 above)
Comment (ItamarB-DI) [reply IO/Ex1.cpp:22]: added in the header

File: IO/Ex1.cpp (hunk 11 above)
Comment (idoa-DI) [reply IO/Ex1.cpp:49]: nevermind . missed the comment in the header

File: IO/Ex1.cpp (hunk 12 above)
Comment (idoa-DI) [reply IO/Ex1.cpp:22]: nevermind . missed the comment in the header\

File: IO/Ex1.hpp [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    FileHandler(FileHandler &other); //CCtor non-const because that tellg/p are modifying 
+    FileHandler(FileHandler &&other) noexcept; //MCtor
+
+    FileHandler& operator=(FileHandler &other); //Copy Assignment non-const because that tellg/p are modifying
+    FileHandler& operator=(FileHandler &&other) noexcept; //Move Assignment
+
+    std::fstream &getFile();
+    static void createFile(const std::string filename);
+
+    ~FileHandler() noexcept = default; //Dtor
```
Comment (idoa-DI) [inline IO/Ex1.hpp]: from the requirements:
```
File stream cleanup should be taken care of at object's destruction
```

File: IO/IO_Ex1_Test.cpp [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+
+    return 0;
+}
+
+void simpleCopyTest() {
+
+    const std::string file_name = "test_file1.txt";
+    std::ios_base::openmode prem = std::ios::in | std::ios::out;
+
+    FileHandler::createFile(file_name);
```
Comment (idoa-DI) [inline IO/IO_Ex1_Test.cpp]: what would happen in a second execution of this test?

File: IO/IO_Ex1_Test.cpp [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+
+    return 0;
+}
+
+void simpleCopyTest() {
+
+    const std::string file_name = "test_file1.txt";
+    std::ios_base::openmode prem = std::ios::in | std::ios::out;
+
+    createFile(file_name);
```
Comment (idoa-DI) [inline IO/IO_Ex1_Test.cpp]: what would happen in a second execution of this test?

File: IO/Ex1.cpp [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+
+std::fstream &FileHandler::getFile() {
+
+    return m_stream;
+}
+
+FileHandler::~FileHandler() noexcept {
+
+    m_stream.flush();
+    m_stream.close();
```
Comment (idoa-DI) [inline IO/Ex1.cpp]: > (destructor)
> [virtual] (implicitly declared)
> 	destructs the basic_fstream and the associated buffer, closes the file
> (virtual public member function)

from https://en.cppreference.com/w/cpp/io/basic_fstream

File: IO/Ex1.cpp [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+    return *this;
+}
+
+//Move Assignment
+FileHandler& FileHandler::operator=(FileHandler &&other) noexcept {
+
+    assert(this != &other);
+
+    if (m_stream.is_open()) {
+        m_stream.close();
```
Comment (idoa-DI) [inline IO/Ex1.cpp]: flush?

File: IO/Ex1.cpp [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+    other.m_permissions = std::ios_base::openmode(0);
+}
+
+//Copy Assignment, non-const because of tellg/p which are modifying 
+FileHandler& FileHandler::operator=(FileHandler &other) {
+
+    assert(this != &other);
+
+    if (m_stream.is_open()) {
+        m_stream.close();
```
Comment (idoa-DI) [inline IO/Ex1.cpp]: flush?

File: IO/IO_Ex1_Test.cpp (hunk 16 above)
Comment (ItamarB-DI) [reply IO/IO_Ex1_Test.cpp]: the createFile just opens the file if it exists, multiple executions shouldn't cause any problems.

File: IO/Ex1.hpp [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+class FileHandler {
+public:
+    explicit FileHandler(const std::string path, std::ios_base::openmode permissions); //Ctor
+    FileHandler(FileHandler &other); //CCtor non-const because that tellg/p are modifying 
+    FileHandler(FileHandler &&other) noexcept; //MCtor
+
+    FileHandler& operator=(FileHandler &other); //Copy Assignment non-const because that tellg/p are modifying
+    FileHandler& operator=(FileHandler &&other) noexcept; //Move Assignment
+
+    void ReOpen(std::ios_base::openmode permissions);//Closes and reopen the file with new permissions
```
Comment (idoa-DI) [inline IO/Ex1.hpp]: use `camelCase` not `PascalCase` in methods

File: IO/Ex1.hpp [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+#include <vector>
+
+
+class FileHandler {
+public:
+    explicit FileHandler(const std::string path, std::ios_base::openmode permissions); //Ctor
+    FileHandler(FileHandler &other); //CCtor non-const because that tellg/p are modifying 
+    FileHandler(FileHandler &&other) noexcept; //MCtor
+
+    FileHandler& operator=(FileHandler &other); //Copy Assignment non-const because that tellg/p are modifying
```
Comment (idoa-DI) [inline IO/Ex1.hpp]: Two spaces before in-line comments and one space between `//` and the comment itself

File: IO/Ex1.hpp [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+    FileHandler& operator=(FileHandler &&other) noexcept; //Move Assignment
+
+    void ReOpen(std::ios_base::openmode permissions);//Closes and reopen the file with new permissions
+    void Write(const std::vector<char> data);
+    std::vector<char> Read(size_t size_to_read);
+    std::uintmax_t Size();
```
Comment (idoa-DI) [inline IO/Ex1.hpp]: What other operations on a file handler should our class provide?

File: IO/Ex1.cpp [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+
+   Close();
+}
+
+
+void FileHandler::Open() {
+
+    m_stream.open(m_path, m_permissions);
+    if (!m_stream.is_open()) {
+        throw std::runtime_error("Failed to open file: " + m_path);
```
Comment (idoa-DI) [inline IO/Ex1.cpp:231]: Can we propagate the reason for the failure?

File: IO/Ex1.cpp [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    m_stream.exceptions(std::ios::failbit | std::ios::badbit);
```
Comment (idoa-DI) [inline IO/Ex1.cpp]: Why?

File: IO/Ex1.hpp (hunk 20 above)
Comment (idoa-DI) [inline IO/Ex1.hpp]: Why `reOpen()` gets only permissions?
What is the usecase for this method?

File: IO/Ex1.cpp [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+    m_stream.seekg(0, std::ios::end);             
+
+    auto eof_idx = m_stream.tellg();
+    if (eof_idx == -1) {
+        throw std::runtime_error("Tell Failed");
+    } 
+
+    m_stream.seekg(m_read_pos);     
+
+    return eof_idx - m_read_pos;
```
Comment (idoa-DI) [inline IO/Ex1.cpp:184]: Is this the size of the file?

File: IO/Ex1.hpp [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+    ~FileHandler() noexcept; //Dtor
+
+private:
+    void Close();
+    void Open();
+
+    std::fstream m_stream;
+    std::string m_path;
+    std::ios_base::openmode m_permissions;
+    std::streampos m_read_pos;
```
Comment (idoa-DI) [inline IO/Ex1.hpp:38]: Why do you need to keep the read position of the file as a member?

File: IO/IO_Ex1_Test.cpp [hunk 27]
```diff
@@ -0,0 +1,169 @@
...
+
+    FileHandler file1(file_name, prem);
+
+    std::fstream &file1_fd = file1.getFile();
```
Comment (idoa-DI) [inline IO/IO_Ex1_Test.cpp]: Didn't you removed the getFile() method?
Please fix your tests so they will compile.

**Review by idoa-DI — CHANGES_REQUESTED** [review]

File: IO/Ex1.cpp (hunk 23 above)
Comment (ItamarB-DI) [reply IO/Ex1.cpp:231]: what do you mean by that doesn't the throw considered as propagate ?

File: IO/Ex1.cpp (hunk 24 above)
Comment (ItamarB-DI) [reply IO/Ex1.cpp]: this makes that errors that set those failure bits throw an exception too ,
which i think it is good so i dont need to check manually after those functions.

File: IO/Ex1.hpp (hunk 20 above)
Comment (ItamarB-DI) [reply IO/Ex1.hpp]: it is for users that opened the file in one way and wants to check the kind of opening.
e.g in one place they only need the read permissions and later on they need the write too .

File: IO/Ex1.cpp (hunk 25 above)
Comment (ItamarB-DI) [reply IO/Ex1.cpp:184]: you are right i updated the name of the method and added a size method.

File: IO/Ex1.hpp (hunk 26 above)
Comment (ItamarB-DI) [reply IO/Ex1.hpp:38]: fstream put and get pointers are the same one,
so i keep track of the get pointer.
like that after write commands if the user reads he gets what he expects

File: IO/IO_Ex1_Test.cpp (hunk 27 above)
Comment (ItamarB-DI) [reply IO/IO_Ex1_Test.cpp]: i removed it and the old tests

File: IO/Ex1.cpp (hunk 23 above)
Comment (yoavs-di) [reply IO/Ex1.cpp:231]: there are many reasons for open to fail. file not exists, bad permission denied etc. please give indication to what happen.

File: IO/Ex1.cpp [hunk 28]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+
+FileHandler::~FileHandler() noexcept {
+
+   close();
+}
+
+
+void FileHandler::open() {
```
Comment (yoavs-di) [inline IO/Ex1.cpp]: open() is a syscall you are overwriting with this naming. please re-name

File: IO/Ex1.cpp [hunk 29]
```diff
@@ -0,0 +1,216 @@
...
+
+    m_stream.open(m_path, m_permissions);
+    if (!m_stream.is_open()) {
+        throw std::runtime_error("Failed to open file: " + m_path);
+    }
+
+    m_stream.exceptions(std::ios::failbit | std::ios::badbit);
+}
+
+void FileHandler::close() {
```
Comment (yoavs-di) [inline IO/Ex1.cpp]: re-name this as well

File: IO/Ex1.cpp [hunk 30]
```diff
@@ -0,0 +1,216 @@
...
+  m_permissions(other.m_permissions),
+  m_read_pos(other.m_read_pos) {
+
+    open(); 
+
+    auto g = other.m_stream.tellg();
+    auto p = other.m_stream.tellp();
+    if (g == -1 || p == -1) {
+        throw std::runtime_error("CCtor Failed to do tellg/p on file: " + m_path);
+    }
```
Comment (yoavs-di) [inline IO/Ex1.cpp]: i see some repetition here.
this check want to answer the question "is the file/stream in a valid state"
are there any other places you want to check this? can you extract to a function?

File: IO/Ex1.cpp [hunk 31]
```diff
@@ -0,0 +1,216 @@
...
+void FileHandler::reOpen(std::ios_base::openmode permissions) {
+
+    close();
+
+    m_permissions = permissions;
+
+    open();
+}
+
+void FileHandler::write(const std::vector<char> data) {
```
Comment (yoavs-di) [inline IO/Ex1.cpp:98]: what happends if another user deleted the file between c`tor and this?
relevant to most api

File: IO/Ex1.cpp [hunk 32]
```diff
@@ -0,0 +1,216 @@
...
+
+    m_stream.write(data.data(), data.size());
+}
+
+std::vector<char> FileHandler::read(size_t size_to_read) {
+    
+    std::vector<char> read_buff(size_to_read, 0);
```
Comment (yoavs-di) [inline IO/Ex1.cpp]: can you contract the vector from iterator?

File: IO/Ex1.cpp
Comment (yoavs-di) [inline IO/Ex1.cpp:1]: seekg and tellg already throws and logs issues, you might want to remove some redundant throws.

Comment (yoavs-di) [inline IO/Ex1.cpp:1]: almost all operations on a file requires the file to be in a valid state, you have implemented checks for that in some but not all function.
can you have one function that guarantee validity?
please use the streams  basic_ios accessors

**Review by yoavs-di — CHANGES_REQUESTED** [review]

File: IO/Ex1.cpp (hunk 28 above)
Comment (ItamarB-DI) [reply IO/Ex1.cpp]: done

File: IO/Ex1.cpp (hunk 29 above)
Comment (ItamarB-DI) [reply IO/Ex1.cpp]: done

File: IO/Ex1.cpp (hunk 31 above)
Comment (ItamarB-DI) [reply IO/Ex1.cpp:98]: open would throw an error

File: IO/Ex1.cpp (hunk 32 above)
Comment (ItamarB-DI) [reply IO/Ex1.cpp]: done

File: IO/Ex1.cpp
Comment (ItamarB-DI) [reply IO/Ex1.cpp:1]: done

File: IO/Ex1.cpp [hunk 33]
```diff
@@ -0,0 +1,244 @@
...
+void FileHandler::reOpen(std::ios_base::openmode permissions) {
+
+    closeFile();
+
+    m_permissions = permissions;
+
+    openFile();
+}
+
+void FileHandler::write(const std::vector<char> data) {
```
Comment (idoa-DI) [inline IO/Ex1.cpp:98]: is data valid?

File: IO/Ex1.cpp [hunk 34]
```diff
@@ -0,0 +1,244 @@
...
+        throw std::runtime_error("File doesn't have permission to read");
+    }
+
+    std::vector<char> read_buff(size_to_read, 0);
+    m_stream.seekg(m_read_pos);
+
+    std::istreambuf_iterator<char> it(m_stream);
+
+    try {
+        std::copy_n(it, size_to_read, read_buff.begin());
```
Comment (idoa-DI) [inline IO/Ex1.cpp:135]: what would happen when it will point to EOF?

File: IO/Ex1.cpp [hunk 35]
```diff
@@ -0,0 +1,244 @@
...
+    try {
+        m_stream.write(data.data(), data.size());
+    } catch (const std::exception &e) {
+        m_stream.clear();
+        throw std::runtime_error(e.what());
+    }
+
+}
+
+std::vector<char> FileHandler::read(size_t size_to_read) {
```
Comment (idoa-DI) [inline IO/Ex1.cpp:117]: if we reach EOF - return the data we read instead of throwing

File: IO/Ex1.cpp [hunk 36]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    m_stream.seekg(off, way);
+
+    auto new_pos = m_stream.tellg();
+
+    m_read_pos = new_pos;
+}
+
+std::uintmax_t FileHandler::sizeUntillEOF() {
```
Comment (idoa-DI) [inline IO/Ex1.cpp]:
```suggestion
std::uintmax_t FileHandler::sizeUntilEOF() {
```

File: IO/Ex1.cpp [hunk 37]
```diff
@@ -0,0 +1,244 @@
...
+    m_stream.seekg(0, std::ios::end);             
+
+    auto eof_idx = m_stream.tellg();
+
+    m_stream.seekg(m_read_pos);     
+
+    return eof_idx - m_read_pos;
+}
+
+std::uintmax_t FileHandler::size() {
```
Comment (idoa-DI) [inline IO/Ex1.cpp:187]: THis logic can be simpler

**Pull request merged**